### PR TITLE
posix: implement `uname`

### LINKED ISSUE #59924: posix: implement `uname()`
There are currently 20 missing POSIX functions with trivial implementations, `uname()` is one of them.

> uname - get the name of the current system
> 
> The uname() function shall store information identifying the current system in the structure pointed to by name.
> 
> The uname() function uses the utsname structure defined in [<sys/utsname.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_utsname.h.html).
> 
> The uname() function shall return a string naming the current system in the character array sysname. Similarly, nodename shall contain the name of this node within an implementation-defined communications network. The arrays release and version shall further identify the operating system. The array machine shall contain a name that identifies the hardware that the system is running on.
> 
> The format of each member is implementation-defined.

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

See RFC #51211 for more info.

### PR DESCRIPTION
**Pull request by ycsin**
Implementation for posix api: `uname`

```
*** Booting Zephyr OS build zephyr-v3.4.0-633-gc38dd60da451 ***

Printing everything in utsname...
sysname[7]: Zephyr
nodename[7]: zephyr
release[9]: 3.4.99
version[61]: zephyr-v3.4.0-633-gc38dd60da451 Jul  6 2023 10:29:50
machine[8]: riscv64


uart:~$ uname -a
Zephyr zephyr 3.4.99 zephyr-v3.4.0-633-gc38dd60da451 Jul  6 2023 10:29:50 riscv64 qemu_riscv64
```

MacOS as a reference:
```
(.venv) ycsin-mba:zephyrproject ycsin$ uname -a
Darwin ycsin-mba 22.5.0 Darwin Kernel Version 22.5.0: Mon Apr 24 20:53:44 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T8103 arm64
```

fixes #59924

- [x] update `doc/services/portability/posix.rst`
- [x] add to the testsuite
- [x] [size_t](https://stackoverflow.com/a/1409631)

### CODE REVIEW

File: include/zephyr/posix/sys/utsname.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+#define ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+
+#include <errno.h>
+#include <zephyr/toolchain/common.h>
+#include <zephyr/net/hostname.h>
+#include "version.h"
```
Comment (cfriedt) [inline include/zephyr/posix/sys/utsname.h:12]: Just a minor nit - I can't remember which RFC it was but we should preemptively adopt @gmarull 's suggestion for canonical header ordering.

```suggestion
#include "version.h"

#include <errno.h>

#include <zephyr/net/hostname.h>
#include <zephyr/toolchain/common.h>
```

File: include/zephyr/posix/sys/utsname.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#define VERSION_BUILD KERNEL_VERSION_STRING
+#endif
+#define VERSION_BUILD_LEN sizeof(VERSION_BUILD)
+
+#define VERSION_NAMELEN (VERSION_BUILD_LEN + DATETIME_LEN)
+
+/* machine */
+#define MACHINE_NAMELEN sizeof(CONFIG_ARCH)
+
+struct utsname {
```
Comment (cfriedt) [inline include/zephyr/posix/sys/utsname.h:50]: Minor comments - it looks as though the array length should be unspecified according to the spec, but if that would cause errors, then no problem.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_utsname.h.html

File: lib/posix/utsname.c [hunk 3]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2023 Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/posix/sys/utsname.h>
+#include <stdio.h>
+#include <string.h>
```
Comment (cfriedt) [inline lib/posix/utsname.c:9]:
```suggestion
#include <stdio.h>
#include <string.h>

#include <zephyr/posix/sys/utsname.h>
```

File: lib/posix/utsname.c [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+int uname(struct utsname *name)
```
Comment (cfriedt) [inline lib/posix/utsname.c:11]: I'm not usually one for adding to the ROM footprint, but it might be significantly simpler to have 1 `static const struct utsname` in this file and then `memcpy()` it to the destination.

It makes sense why we have the array sizes now, so please disregard my earlier Q.

**Review by cfriedt — CHANGES_REQUESTED** [review]
LGTM - just a couple of nits and suggestions.

The sample is good, but it would also be good to add to the testsuite.

One tricky area in particular is the version string.

We actually do want to be fairly consistent here with e.g. the boot banner.

So it would make sense to detect and modify the string such that we can display
* a tagged release (e.g. `v3.5.0`)
* something commit-ish (IIRC git short ref plus epoch time? I can't remember)
* whether the build is dirty (Linux uses a `+`, not sure how Zephyr does it)

Of possible, can you please also update `doc/services/portability/posix.rst`?

Great work! Thanks @ycsin 🚀🪁👍🍍

**Comment by cfriedt** [comment]
Some compliance fixes..

**Comment by cfriedt** [comment]
IIRC, we cannot yet build for `native_posix`. I would suggest filtering `CONFOG_ARCH_POSIX` in your sample

File: lib/posix/utsname.c (hunk 4 above)
Comment (ycsin) [reply lib/posix/utsname.c:11]: do we need a mechanism to detect if hostname is changed if `net_hostname_set_postfix` is enabled, and update that static copy accordingly?

**Comment by ycsin** [comment]
> We actually do want to be fairly consistent here with e.g. the boot banner.

The `version` string is the same as the [boot banner](https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/banner.c#L19) (I just copied & renamed)

> * a tagged release (e.g. `v3.5.0`)
> * something commit-ish (IIRC git short ref plus epoch time? I can't remember)

Should be covered by:
```
3.4.99 zephyr-v3.4.0-632-gfed4ab1928ba Jul  4 2023 19:01:24
```

Not epoch time though..

> whether the build is dirty (Linux uses a `+`, not sure how Zephyr does it)

Looks like something to be done in the build system, not sure if @tejlmand has any idea?
I vaguely rmb that we used to have this somewhere..?

File: lib/posix/utsname.c [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/posix/sys/utsname.h>
+
+int uname(struct utsname *name)
+{
+	if (name == NULL) {
+		return -1;
+		errno = EINVAL;
```
Comment (ycsin) [inline lib/posix/utsname.c:16]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html:

> Upon successful completion, a non-negative value shall be returned. Otherwise, -1 shall be returned and errno set to indicate the error.

File: lib/posix/utsname.c [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/posix/sys/utsname.h>
+
+int uname(struct utsname *name)
+{
+	if (name == NULL) {
```
Comment (cfriedt) [inline lib/posix/utsname.c:14]: I wouldn't even bother with the runtime check.  "No errors defined".
https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

File: lib/posix/utsname.c (hunk 5 above)
Comment (cfriedt) [reply lib/posix/utsname.c:16]: I wouldn't even bother with the runtime check.  "No errors defined".
https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

File: lib/posix/utsname.c (hunk 4 above)
Comment (cfriedt) [reply lib/posix/utsname.c:12]: Ok then, `memcpy()` + `snprintf()` for the hostname.

File: lib/posix/utsname.c (hunk 5 above)
Comment (ycsin) [reply lib/posix/utsname.c]: Replaced that runtime check with `__ASSERT_NO_MSG`

**Comment by tejlmand** [comment]
> > whether the build is dirty (Linux uses a `+`, not sure how Zephyr does it)
> 
> Looks like something to be done in the build system, not sure if @tejlmand has any idea? I vaguely rmb that we used to have this somewhere..?

Adding `-dirty` (or anything indicating a dirty build) has been rejected in the past, ref: https://github.com/zephyrproject-rtos/zephyr/pull/13608#pullrequestreview-206725920

But the use-case / reason for not having `-dirty` can be handled manually by those who has special needs, as described here:
https://github.com/zephyrproject-rtos/zephyr/issues/39503#issuecomment-961686122

As people has various ways of working / verifying patches, then I think a wider audience should be invited to discuss if we want to have `-dirty` (or similar) appended to short sha.
The implementation itself is very trivial, so just a matter of making a decision.

File: lib/posix/utsname.c [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+#if defined(__DATE__) && defined(__TIME__)
+	snprintf(name->version, sizeof(name->version), "%s %s %s", VERSION_BUILD, __DATE__,
+		 __TIME__);
+#else
+	strncpy(name->version, VERSION_BUILD, sizeof(name->version));
+#endif
+	strncpy(name->machine, CONFIG_ARCH, sizeof(name->machine));
+
+	return 0;
+}
```
Comment (cfriedt) [inline lib/posix/utsname.c:29]:
```suggestion
#include <string.h>

#include <zephyr/kernel.h>
#include <zephyr/posix/sys/utsname.h>

#ifdef CONFIG_NET_HOSTNAME_ENABLE
#define UTSNAME_NODENAME CONFIG_NET_HOSTNAME
#else
#define UTSNAME_NODENAME "zephyr"
#endif

#if defined(__DATE__) && defined(__TIME__)
#define UTSNAME_VERSION(_ver) _ver " " __DATE__ " " __TIME__
#else
#define UTSNAME_VERSION(_ver) _ver
#endif

#define UTSNAME_INITIALIZER(_sys, _node, _rel, _ver, _mach)                                        \
	{                                                                                          \
		.sysname = _sys, .nodename = _node, .release = _rel,                               \
		.version = UTSNAME_VERSION(_ver), .machine = _mach,                                \
	}

static const struct utsname z_name = UTSNAME_INITIALIZER(
	"Zephyr", UTSNAME_NODENAME, KERNEL_VERSION_STRING, VERSION_BUILD, CONFIG_ARCH);

int uname(struct utsname *name)
{
	memcpy(name, &z_name, sizeof(*name));
	if (IS_ENABLED(CONFIG_NET_HOSTNAME_ENABLE)) {
		strncpy(name->nodename, net_hostname_get(), sizeof(name->nodename));
	}

	return 0;
}
```

File: include/zephyr/posix/sys/utsname.h [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+#define ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+
+#include "version.h"
+
+#include <errno.h>
+#include <stdint.h>
+
+#include <zephyr/net/hostname.h>
+#include <zephyr/toolchain/common.h>
+#include <zephyr/net/hostname.h>
```
Comment (cfriedt) [inline include/zephyr/posix/sys/utsname.h:16]:
```suggestion
#include <zephyr/kernel.h>
```

File: include/zephyr/posix/sys/utsname.h [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+ * be formatted in utsname.c as:
+ *     "%s %s %s", VERSION_BUILD, __DATE__, __TIME__
+ * the '\0' of VERSION_BUILD_LEN & __DATE__ will be used for the space that
+ * follows, '\0' of __TIME__ will be used for the \0 of the formatted VERSION
+ * string. In the end the size of the array should fit the string snugly.
+ */
+#define VERSION_NAMELEN (VERSION_BUILD_LEN + DATETIME_LEN)
+
+/* machine */
+#define MACHINE_NAMELEN sizeof(CONFIG_ARCH)
```
Comment (cfriedt) [inline include/zephyr/posix/sys/utsname.h:60]: While I appreciate that this is all very precise, I kind of think we would be better off keeping it simple for now.

E.g. add a Kconfig integer option `CONFIG_POSIX_UTSNAME_LENGTH`, and use
```
#ifdef CONFIG_POSIX_UTSNAME
#define _UTSNAME_LENGTH CONFIG_POSIX_UTSNAME_LENGTH
#else
#define _UTSNAME_LENGTH sizeof("Zephyr")
#endif
```

https://man7.org/linux/man-pages/man2/uname.2.html

Inside of the `utsname.c` (which may be better to name `uname.c`), we can have several compile-time `BUILD_ASSERT()` statements. E.g.

```cpp
BUILD_ASSERT(sizeof(z_name.sysname) >= sizeof("Zephyr"));
BUILD_ASSERT(sizeof(z_name.release) >= sizeof(KERNEL_VERSION_STRING));
BUILD_ASSERT(sizeof(z_name.version) >= sizeof(UTSNAME_VERSION(VERSION_BUILD));
BUILD_ASSERT(sizeof(z_name.machine) >= sizeof(CONFIG_ARCH));

```

File: include/zephyr/posix/sys/utsname.h [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+struct utsname {
+	char sysname[SYS_NAMELEN];
+	char nodename[NODE_NAMELEN];
+	char release[RELEASE_NAMELEN];
+	char version[VERSION_NAMELEN];
+	char machine[MACHINE_NAMELEN];
```
Comment (cfriedt) [inline include/zephyr/posix/sys/utsname.h:67]:
```suggestion
	char sysname[_UTSNAME_LENGTH];
	char nodename[_UTSNAME_LENGTH];
	char release[_UTSNAME_LENGTH];
	char version[_UTSNAME_LENGTH];
	char machine[_UTSNAME_LENGTH];
```

File: include/zephyr/posix/sys/utsname.h [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+int uname(struct utsname *name);
+
+#undef SYS_NAMELEN
+#undef NODE_NAMELEN
+#undef RELEASE_NAMELEN
+#undef VERSION_NAMELEN
+#undef MACHINE_NAMELEN
+#undef DATETIME_LEN
+#undef VERSION_BUILD_LEN
+
```
Comment (cfriedt) [inline include/zephyr/posix/sys/utsname.h:79]:
```suggestion
```

File: lib/posix/Kconfig [hunk 12]
```diff
@@ -182,3 +182,10 @@ config FNMATCH
 	help
 	  Match filenames using the the fnmatch function. For example, the pattern
 	  "*.c" matches the filename "hello.c".
+
+config POSIX_UNAME
+	bool "Support for uname"
+	default y if POSIX_API
+	help
+	  The uname() function shall store information identifying the current
+	  system in the structure pointed to by name.
```
Comment (cfriedt) [inline lib/posix/Kconfig:191]: Would be good to also add `config POSIX_UTSNAME_LENGTH`, see above

File: samples/posix/uname/src/main.c [hunk 13]
```diff
@@ -0,0 +1,186 @@
+/*
```
Comment (cfriedt) [inline samples/posix/uname/src/main.c:1]: This is great - we may actually want to move this into `lib/posix` at some point. Thanks for adding it :-)

**Comment by cfriedt** [comment]
An alternative internal thing for `uname.c` is below, where we don't copy constant string data into an array.

With this setup, it would make sense to use `strncpy()` for each field.
```cpp
struct utsname_internal {
	const char *const sysname;
	const char *const nodename;
	const char *const release;
	const char *const version;
	const char *const machine;
};
```

**Comment by cfriedt** [comment]
> Not epoch time though..

`__TIME_UNIX__` might work - I think it's epoch time

File: include/zephyr/posix/sys/utsname.h (hunk 8 above)
Comment (ycsin) [reply include/zephyr/posix/sys/utsname.h]: moved all #include into the source file

**Review by cfriedt — DISMISSED** [review]

File: include/zephyr/posix/sys/utsname.h [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct utsname {
+	char sysname[sizeof("Zephyr")];
+	char nodename[CONFIG_POSIX_UNAME_NODENAME_LEN];
+	char release[sizeof("99.99.99")];
+	char version[CONFIG_POSIX_UNAME_VERSION_LEN];
```
Comment (keith-packard) [inline include/zephyr/posix/sys/utsname.h:17]: I think the size of this field can also be computed -- just need to insert the computation of the version string into this header. One less config value to deal with seems like a nice thing.

File: lib/posix/Kconfig [hunk 15]
```diff
@@ -182,3 +182,26 @@ config FNMATCH
...
+
+if POSIX_UNAME
+config POSIX_UNAME_VERSION_LEN
+	int "Buffer size for version"
+	default 60
+	help
+	  Size of the version char array in the utsname struct.
+
+config POSIX_UNAME_NODENAME_LEN
```
Comment (keith-packard) [inline lib/posix/Kconfig:200]: I looked at the Linux version of this code and it defines the maximum string length and then adds one when declaring the array to account for the null terminator. Seems like a good idea. Would mean changing the doc strings here to talk about 'maximum length of the node name'. (the doc string here was also copied from the version one above).

File: lib/posix/uname.c [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+BUILD_ASSERT(sizeof(z_name.sysname) >= sizeof("Zephyr"));
+BUILD_ASSERT(sizeof(z_name.release) >= sizeof(KERNEL_VERSION_STRING));
+BUILD_ASSERT(sizeof(z_name.version) >= sizeof(UTSNAME_VERSION(VERSION_BUILD)));
+BUILD_ASSERT(sizeof(z_name.machine) >= sizeof(CONFIG_ARCH));
+
+int uname(struct utsname *name)
+{
+	memcpy(name, &z_name, sizeof(*name));
+	if (IS_ENABLED(CONFIG_NET_HOSTNAME_ENABLE)) {
+		strncpy(name->nodename, net_hostname_get(), sizeof(name->nodename));
```
Comment (keith-packard) [inline lib/posix/uname.c:49]: strncpy doesn't ensure the result is null terminated.

**Review by keith-packard — CHANGES_REQUESTED** [review]
the strncpy usage needs to be fixed, otherwise my comments are just suggestions.

File: include/zephyr/posix/sys/utsname.h (hunk 9 above)
Comment (ycsin) [reply include/zephyr/posix/sys/utsname.h]: @keith-packard if we remove the `CONFIG_POSIX_UNAME_VERSION_LEN`, then we will have to reinstate the complicated part of the computation here

**Comment by ycsin** [comment]
> Adding `-dirty` (or anything indicating a dirty build) has been rejected in the past

how about having a separate define to indicate this in "version.h", rather than appending to the hash directly?

File: lib/posix/uname.c (hunk 16 above)
Comment (ycsin) [reply lib/posix/uname.c:49]: added `\0` to the end of the buffer

**Comment by tejlmand** [comment]
> how about having a separate define to indicate this in "version.h", rather than appending to the hash directly?

Did you read the links ?
Especially: https://github.com/zephyrproject-rtos/zephyr/pull/13608#pullrequestreview-206725920 .

Even if placing this info in a separate define to use in `version.h` then it will impact the hex file when the repo goes from clean --> dirty (or other way around).

So if this is to be changed, then please open a dedicated issue for discussing, and not in this PR.

File: samples/posix/uname/src/main.c [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+	ret = uname(&info);
+	if (ret < 0) {
+		shell_error(sh, "cannot get system name");
+		return -1;
+	}
+
+	/* Process each option */
+
+	first = true;
+	for (i = 0; set != 0; i++) {
```
Comment (cfriedt) [inline samples/posix/uname/src/main.c:138]: This can be done without a for loop so that it uses symbolic values instead of magic numbers.

e.g. `if (set & UNAME_VERSION)`

**Review by cfriedt — CHANGES_REQUESTED** [review]
One minor change request

**Comment by cfriedt** [comment]
> So if this is to be changed, then please open a dedicated issue for discussing, and not in this PR.

@tejlmand - I believe @ycsin's changes have addressed your concerns, and with a couple of small changes, I'm ready to approve. Can you take a look again to ensure that your concerns were addressed?

File: samples/posix/uname/src/main.c [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+	printk("sysname[%zu]: %s\n", sizeof(info.sysname), info.sysname);
+	printk("nodename[%zu]: %s\n", sizeof(info.nodename), info.nodename);
+	printk("release[%zu]: %s\n", sizeof(info.release), info.release);
+	printk("version[%zu]: %s\n", sizeof(info.version), info.version);
+	printk("machine[%zu]: %s\n", sizeof(info.machine), info.machine);
+
+	return 0;
+}
+
+#define UNAME_KERNEL   (1 << 0)
```
Comment (cfriedt) [inline samples/posix/uname/src/main.c:30]: @ycsin - sorry - I missed this (was reviewing on mobile).

Please use `BIT(n)` here

File: tests/posix/common/src/uname.c [hunk 19]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2023 meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/ztest.h>
+#include <sys/utsname.h>
+
+ZTEST(posix_apis, test_posix_uname)
```
Comment (cfriedt) [inline tests/posix/common/src/uname.c:10]: @ycsin - one last thing (sorry again). Let's shorten this to `test_uname`).

File: include/zephyr/posix/sys/utsname.h (hunk 14 above)
Comment (ycsin) [reply include/zephyr/posix/sys/utsname.h]: We can actually compute everything using preprocessor defines here, at the expense of including 2 more headers & some macro magic then looks a bit complicated

**Review by cfriedt — APPROVED** [review]

**Comment by cfriedt** [comment]
@keith-packard - does this one look OK to you now?

File: include/zephyr/posix/sys/utsname.h (hunk 14 above)
Comment (cfriedt) [reply include/zephyr/posix/sys/utsname.h]: The version field can be fixed size as well, technically, too, if we omit the date and time, and it's compatible with the banner version as well.

```cpp
char version[sizeof("zephyr-v99.99.99-9999-g0123456789ab")];
```

Ideally we could just include `<version.h>`, which is generated at build time and use the exact `#define`s that appear there.

I'm hesitant to do that, just because it really isn't referenced outside of `banner.c`, which gives me the impression that it's private somehow.

If that makes life easier though, by all means, just 
```cpp
struct utsname {
	char sysname[sizeof("Zephyr")];
	char nodename[CONFIG_POSIX_UNAME_NODENAME_LEN + 1];
	char release[sizeof(KERNEL_VERSION_STRING)];
	char version[sizeof(STRINGIFY(BUILD_VERSION))];
	char machine[sizeof(CONFIG_ARCH)];
};
```

And honestly, the `__DATE__` and `__TIME__` macros are probably not *super* critical.

**Comment by cfriedt** [comment]
@ycsin - I'm OK getting 1 more approval for this and merging if you wouldn't mind a follow-up PR for namespacing `<version.h>` as `<zephyr/kernel/banner_version.h>` or something along those lines.

**Review by keith-packard — APPROVED** [review]
Yeah, given the concerns about mixing too many bits into the application namespace, having compile-time checks that make sure the sizes are big enough seems good enough here.

File: include/zephyr/posix/sys/utsname.h (hunk 14 above)
Comment (marc-hb) [reply include/zephyr/posix/sys/utsname.h]: `__DATE__` and `__TIME__` break build reproducibility and should not be enabled by default (not sure what is the current situation)
